### PR TITLE
[Rest Server] Fix Docker container OOM issue

### PR DESCRIPTION
**Pull request by abuccts**
Kill Docker container when yarn container exit.

### CODE REVIEW

**Comment by abuccts** [comment]
Should we set [`--oom-kill-disable` flag](https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources) or not?

And it's better to figure out Docker container OOM issue after changing container log directory.

File: rest-server/src/templates/yarnContainerScript.mustache [hunk 1]
```diff
@@ -31,16 +34,14 @@ function exit_handler()
   exec 13>&-
   hdfs dfs -put /tmp/pai_yarncontainer_$CONTAINER_ID.log \
     {{{ hdfsUri }}}/Container/{{{ jobData.userName }}}/{{{ jobData.jobName }}}/log/
+  kill -9 $(docker inspect --format="{{.State.Pid}}" $docker_name)
```
Comment (fanyangCS) [inline rest-server/src/templates/yarnContainerScript.mustache]: >docker inspect [](start = 12, length = 14)

actually, when a docker container runs into OOM and set oom-kill-disable, docker inspect will hang indefinitely

Comment (fanyangCS) [inline rest-server/src/templates/yarnContainerScript.mustache]: >$docker_name [](start = 53, length = 12)

I hope it does not return 0 if the docker_name doesn't exist.

Comment (abuccts) [reply rest-server/src/templates/yarnContainerScript.mustache]: removed

Comment (abuccts) [reply rest-server/src/templates/yarnContainerScript.mustache]: It won't.
- If there's no error in trap handler, the exit code is the same as the last command before trap handler.
- If there's an error in trap handler, the exit code should be the same as that.

Comment (fanyangCS) [reply rest-server/src/templates/yarnContainerScript.mustache]: I am not talking about exit code. I am wondering What is the result of "docker inspect --format="{{.State.Pid}}" $docker_name" when $docker_name does *not* exist (i.e., already exited)

---
In reply to: [195936120](https://github.com/Microsoft/pai/pull/735#discussion_r195936120) [](ancestors = 195936120)

**Pull request merged**